### PR TITLE
ANCM Shim/RequestHandler Refactor

### PR DESCRIPTION
**Pull request by jkotalik**
Will be commenting throughout the PR with updates/description.

Items that still need to be cleaned up (will make issues on these):
- Logging. Transfering to use a new ETW logging style.
- Verification. 
- Updating tests to react to shim/requesthandler split.
- Any TODOs.

Until all bits are stabalized, we will not update the version of ANCM in Universe.

### CODE REVIEW

File: AspNetCoreModule.sln [hunk 1]
```diff
@@ -34,18 +39,28 @@ Global
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{439824F9-1455-4CC4-BD79-B44FA0A16552}.Debug|Any CPU.ActiveCfg = Debug|Win32
-		{439824F9-1455-4CC4-BD79-B44FA0A16552}.Debug|Win32.ActiveCfg = Release|x64
-		{439824F9-1455-4CC4-BD79-B44FA0A16552}.Debug|Win32.Build.0 = Release|x64
+		{439824F9-1455-4CC4-BD79-B44FA0A16552}.Debug|Win32.ActiveCfg = Debug|Win32
```
Comment (jkotalik) [inline AspNetCoreModule.sln]: Update these build configs to all be the same.

File: src/AspNetCore/Inc/applicationinfo.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (jkotalik) [inline src/AspNetCore/Inc/applicationinfo.h:2]: This is effectively a record for what applications are currently created.

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    APPLICATION* pApplication = NULL;
+    STACK_STRU(struFileName, 300);  // >MAX_PATH
+    STRU        hostFxrDllLocation;
+
+    if (m_pApplication != NULL)
+    {
+        goto Finished;
+    }
+
+    FindRequestHandlerAssembly();
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationinfo.cpp]: The requesthandler.dll will have two functions exposed on it; CreateApplication (on a per application basis) and CreateRequestHandler (per request basis). Calling FindRequestHandlerAssembly will call into hostfxr to find it.

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    {
+        AcquireSRWLockExclusive(&g_srwLock);
+        fLocked = TRUE;
+        if (g_fAspnetcoreRHAssemblyLoaded)
+        {
+            goto Finished;
+        }
+
+        // load assembly and create the application
+        if (m_pConfiguration->QueryHostingModel() == APP_HOSTING_MODEL::HOSTING_IN_PROCESS)
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationinfo.cpp:223]: This logic will change based on the changes to hostfxr. The logic (for both in proc and out of proc) will be:
if hostfxrpath exists and hostfxr has the ability to find the requesthandler
   use hostfxr to find the request handler
otherwise if we are running out of process
   use the global requesthandler

Another issue is bitness for inprocess and out of process. If we are running a 32-bit worker process and find a 64 bit hostfxr, out of process can still work, but in-process cannot. However, if we want out-of-process to be loaded the same way we do in process, we would need a hostfxr.dll to be the same bitness as the worker process. We will need to discuss the finer details here.

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    {
+        ReleaseSRWLockExclusive(&g_srwLock);
+    }
+    return hr;
+}
+
+HRESULT
+APPLICATION_INFO::FindNativeAssemblyFromInetsrv(STRU* struFilename)
+{
+    HRESULT hr = S_OK;
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationinfo.cpp]: Eventually will only be used for backwards compatibility.

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 6]
```diff
@@ -128,21 +125,27 @@ APPLICATION_MANAGER::GetApplication(
...
+    if (FAILED(hr))
+    {
         if (fDuplicatedInProcessApp)
         {
             if (SUCCEEDED(strEventMsg.SafeSnwprintf(
                 ASPNETCORE_EVENT_DUPLICATED_INPROCESS_APP_MSG,
                 pszApplicationId)))
             {
-                apsz[0] = strEventMsg.QueryStr();
+                /*apsz[0] = strEventMsg.QueryStr();
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationmanager.cxx:156]: We will be updating logging in a separate PR.

File: src/AspNetCore/Src/inprocessstoredcontext.cxx [hunk 7]
```diff
@@ -1,113 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the MIT License. See License.txt in the project root for license information.
-
-#include "precomp.hxx"
-
-IN_PROCESS_STORED_CONTEXT::IN_PROCESS_STORED_CONTEXT(
```
Comment (jkotalik) [inline src/AspNetCore/Src/inprocessstoredcontext.cxx:6]: Using IN_PROCESS_HANDLER instead of stored context.

File: src/AspNetCore/Src/proxymodule.cxx [hunk 8]
```diff
@@ -41,87 +41,139 @@ Return value:
...
+        retVal = RQ_NOTIFICATION_FINISH_REQUEST;
+        goto Finished;
+    }
+    else
+    {
+        // make sure assmebly is loaded and application is created
+        APPLICATION* pApplication = NULL;
+        STACK_STRU(struFileName, 256);
+
+        hr = pApplicationInfo->EnsureApplicationCreated();
```
Comment (jkotalik) [inline src/AspNetCore/Src/proxymodule.cxx]: EnsureApplicationCreated will load the request handler if need be.

File: src/CommonLib/CommonLib.vcxproj [hunk 9]
```diff
@@ -0,0 +1,204 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jkotalik) [inline src/CommonLib/CommonLib.vcxproj:1]: Common lib is for any shared components between AspNetCore and the RequestHandler projects.

**Review by jkotalik — COMMENTED** [review]
A bit of introduction for some of the inprocess stuff.

**Comment by pan-wang** [comment]
refactor ANCM to support In-process
